jlm/llvm/opt/CommonNodeElimination.cpp

```diff
@@ -0,0 +1,1017 @@
+  if (node.noutputs() == 0)
+    return &node;
+
+  const auto & congruenceSets = context.getCongruenceSets();
+
+  // Check the congruence set of the first output
+  const auto output0Set = congruenceSets.tryGetSetFor(*node.output(0));
+  // If the output has not gotten a congruence set yet, the node has yet to be marked
+  if (output0Set == CongruenceSets::NoCongruenceSet)
+    return nullptr;
+
+  // Structural nodes can have outputs that are congruent with completely different nodes
+  // due to invariant values. Structural nodes are always considered their own leaders.
+  if (dynamic_cast<const rvsdg::StructuralNode *>(&node))
+    return &node;
+
+  // A simple node can only be congruent with other simple nodes
+  const auto & output0Leader = congruenceSets.getLeader(output0Set);
+  const auto & leaderNode = rvsdg::AssertGetOwnerNode<rvsdg::SimpleNode>(output0Leader);
+  JLM_ASSERT(leaderNode.noutputs() == node.noutputs());
+  return &leaderNode;
+}
+
+using TopNodeLeaderList = std::vector<const rvsdg::Node *>;
+
+/**
+ * Function for marking the given top node \p node, marking it congruent with a previously marked
+ * top node if their operations are identical. Otherwise it becomes a leader.
+ * Maintains a separate list of leader top nodes in the region.
+ * @param node the top node to mark.
+ * @param leaders the list of all leader top nodes in the region.
+ * @param context the current context of the marking phase.
+ */
+static void
+markSimpleTopNode(
+    const rvsdg::SimpleNode & node,
+    TopNodeLeaderList & leaders,
+    CommonNodeElimination::Context & context)
+{
+  // If the node has already been marked, check if it is still congruent with its leader
+  if (const auto existingLeaderNode = tryGetLeaderNode(node, context))
+  {
+    // We are our own leader, nothing to check
+    if (existingLeaderNode == &node)
+    {
+      leaders.push_back(&node);
+      return;
+    }
+
+    // Check if we are still congruent with the leader
+    if (checkNodesCongruent(*existingLeaderNode, node, context))
+    {
+      return;
+    }
+
+    // This node is no longer congruent with its leader, continue looking for a new leader
+  }
+
+  // TODO: Use some sort of hashing to make this not O(n * m)
+  // where n is the number of outputs and m is the number of congruence sets
+
+  // Check if the node is congruent with any existing leader in the TopNodeLeaderList
+  for (auto leader : leaders)
+  {
+    if (checkNodesCongruent(node, *leader, context))
+    {
+      markNodesAsCongruent(*leader, node, context);
+      return;
+    }
+  }
+
+  // No existing leader top node found, create new congruence sets for each of node's outputs
+  markNodeAsLeader(node, context);
+  leaders.push_back(&node);
+}
+
+/**
+ * Places the given \p node in a congruence set, based on its operation and inputs.
+ * If the node already has a congruence set, it gets double-checked against the leader.
+ * The node must not be a top node. @see markTopNode.
```

Comment:
markTopNode does not exist. I think you meant markSimpleTopNode.

I suspect the reason why doxygen CI does not pick up on it is because that the docoumentation in cpp files is just removed/ignored.